### PR TITLE
Add ScalersCache to reuse scalers unless they need changing

### LINKED ISSUE #1121: Should scalers reuse opened connection/clients?
I have been thinking about this for some time. Currently [we are creating a new Scaler](https://github.com/kedacore/keda/blob/9a904ccc6a3077323ea1582eddaabff56554498b/pkg/scaling/scale_handler.go#L366) (and creating a client and opening a connection) in every cycle after in the Scale Loop (after `pollingInterval` has passed) and we are doing the same in the Metrics Server when HPA is requesting the metrics.

This has some benefits, eg. if the `pollingInterval` is long, we don't need to keep the opened conection, use the memory resources etc. And if I am not mistaken, then ENV variables on the Target Deployment are evaluated in each cycle.

On the other hand, if the `pollingInterval` is not that long, we keep opening and closing the clients very often. This is not ideal from the performance perspective and we can as well flood the Event Source (eg. Kafka Broker) with many opened and closed connections.

What if we try to create the Scaler just once, and keep the reference on the Scaler (and if there is an error while accessing this scaler, we can recreate again). We should do it on both sides (in the Operator's Scale Loop and in the Metrics server).

Are there any downsides or problems? Or any other ideas?

### PR DESCRIPTION
**Pull request by ahmelsayed**
Closes #1121

_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [N/A] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [N/A] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated

Fixes #1121

### CODE REVIEW

**Comment by zroubalik** [comment]
> That's a good point @zroubalik. I'll just add a feature env var that will invalidate the cache right after every check for now (effectively closing the scaler to have the same old behavior) in case the new behavior impacts someone. How does that sound?

So you propose to introduce global env var that will affect all ScaledObject or do per ScaledObjects setting? I am fine with both approaches, or we can introduce the per ScaledObject option later, if there's a demand.

**Comment by zroubalik** [comment]
@ahmelsayed any update on this please? I'd like to start working on https://github.com/kedacore/keda/issues/2156 to have it for the upcoming release. The best thing would be to base it on your code (the controller part in Metric Server in particular, etc), so don't want start on this until this is merged, to avoid complex rebases 😄 Thx :)

**Comment by ahmelsayed** [comment]
Sorry for the delay, I just rebased the PR, and I believe all the feedback has been addressed? (the conversation isn't easy to follow on github ui)

**Comment by zroubalik** [comment]
> /run-e2e

There is an intermittent failure in `scalers/azure-queue-trigger-auth.test.ts`, the rest seem to be ok 🎉

**Review by JorTurFer — APPROVED** [review]
LGTM!

**Comment by zroubalik** [comment]
@ahmelsayed could you please update the changelog with this contribution?

File: pkg/scaling/cache/scalers_cache.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+		var metricValue int64
+
+		for _, m := range metrics {
+			if m.MetricName == "queueLength" {
+				metricValue, _ = m.Value.AsInt64()
+				queueLength += metricValue
+			}
+		}
+		scalerLogger.V(1).Info("Scaler Metric value", "isTriggerActive", isTriggerActive, "queueLength", queueLength, "targetAverageValue", targetAverageValue)
```
[the remark this answers is not in the document]
Comment (arschles) [reply pkg/scaling/cache/scalers_cache.go:285]: I'm happy to do the log in a follow-up to this PR, in the interest of getting this merged. @ahmelsayed let me know if (a) you think logging is an adequate strategy and (b) you'd like me to do the follow-up

**Review by arschles — APPROVED** [review]
@ahmelsayed sorry I haven't been back to this PR in some time. I left one comment regarding a log line (a nit-pick, really). I'm happy to do it in a follow-up if you think it'd be a good idea. Let me know.

Regardless, LGTM

File: pkg/scaling/cache/scalers_cache.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahmelsayed) [reply pkg/scaling/cache/scalers_cache.go:285]: Yes please @arschles log a follow up item. I think logging sounds like a fine strategy here. Please feel free to take care of it if you'd like :)

**Pull request merged**